Makefile

```diff
@@ -57,8 +57,17 @@ crossdev:
 	cd documents$(S)release$(S)crossdev ; $(CDEL) crossdev.zip ; zip -r -q crossdev.zip *
 
 samples:
-	zip -r -j -q release$(S)samples.zip basic$(S)code basic$(S)images$(S)test$(S)test.gfx basic$(S)images$(S)graphics.gfx
-	zip -d -q release$(S)samples.zip *.tass 
+	# copy BASIC examples into examples/ tree
+	ifeq ($(OS),Windows_NT)
+	     ROBOCOPY basic\code\ examples\basic\ /XF *.tass
+	else
+	     cp -a basic/code/ examples/BASIC/
+	     find -type f -name '*.tass' --delete
+	endif
+
+	# generate examples tarball
+	zip --recurse-paths --quiet samples.zip examples
```

Comment:
I think this samples.zip is in the wrong path to get collected by the release pipeline. It should be in release/samples.zip but I think this will be in /samples.zip.

Reply:
ok i intended to discuss that - notice that the specific directory is unimportant - both inputs (samples.zip and crossdev.zip) are deleted imediatately after re-packing them into neo6502.zip - i suppose it is done that way due to the deficiencies of the `zip` program - that -j option is awful though - i would definitely avoid that

the Makefile would be much cleaner (easier to grok) if the intermediate sourceballs were on the top-level directory - notice how the the crossdev: make target needs to delete 'crossdev.zip' (an artifact of a previous build?) before generating it, to avoid adding it to itself - that is something of a design smell - that build artifact could have been deleted by `make clean`, and anyways not be in the directory of it's own sources in the first place

but to take that a step further, its not obvious why the release file contains samples.zip and crossdev.zip, instead of their expanded contents - after looking into them, it becomes obvious that the the contents of 'crossdev.zip' are actually just more application code examples, nothing different than the contents of 'samples.zip' - so even if there is a rationale for having those two zip files inside the release zip file, i dont know what would be the rationale for having both instead of one - it seems to me that the files under documents/release/crossdev/ should just be moved to examples/BASIC/ with the others, eliminating the separate and extra 'crossdev:' and 'samples:' make targets

next, as with 'crossdev.zip', the emulator Makefiles need to delete 'neolinux.zip' and 'neowin.zip' to avoid adding them to themselves - adding the contents of those intermediate archives directly to neo6502.zip would simplify the process further - noting that 'neolinux.zip' and 'neowin.zip' contain only one file each, the rationale is yet more mysterious - why not simply put them in neo6502.zip as ready-to-use binaries?

so i added a new commit which does all of that

Reply:
I've no problem with them being unexpanded. I think it's a hangover from the original rather chaotic system I came up with :)

Reply:
i agree, it is not a problem - i just cant imagine what is the motivation to put archives inside other archives, especially if they contain only one additional file; so i thought i should shine light on those, to determine if those were just overlooked, or maybe a work in progress which will makes sense as separate entities later - otherwise, if the extra archives are not related to each other and the parent container, they would be better distributed individually, so that the users only need to download and explode the ones they want (eg: only the examples) - if they are all related, then why make the user extract so many? - it just seems like an extra complication, or perhaps some forgotten cruft, which i always try to notice and try to reduce - i am a big fan of simplicity - i understand that you use arch; so you probably are also

in this case, you can see in Makefile, after i simplified things by eliminating the unnecessary and redundant 'crossdev.zip' and 'samples.zip', all three of the zipfile:, crossdev:, and samples: targets collapsed to four easy to grok LOC - it simply adds files to _the_ one archive, sorted by directory - to me, a change of that kind is very satisfying - i would take that approach further, and sort all of the remaining files and extra zip files into directories, rather than everything being unstructured in the same top-level directory (avoid the `zip -j` option)

another factor i would note, is that what the zipfile: target does, is essentially what is normally a 'release:' make target (not to be confused with a "Release" vs "Debug" build target, ala: windows) - when a release: target is present in a Makefile, it is simply a convenience - it's work could be an ordinary script; because all steps will happen each time it is run, unlike what `make` normally does (it sometimes does nothing) - the 'release' procedure is normally not part of the standard `make` procedure; because it only needs to be run by the upstream, and only occasionally - having it as part of the standard build (which every downstream will run every time) is a case of YAGNI (downstreams are not going to need it)

these extra complexities are all relatively minor o/c - i would not want to spend too much time debating them - i only wanted you to see my KISS-minded approach; and i can do more of the same if you like

i realize that i am going on "long-winded", and i am definitely (unashamedly) nit-picking - i tend to approach a code-base holistically; and always prioritize simplicity and tidiness over progress - for example, the use of tabs/white-space in this code-base is hideous, if you dont mind me saying so :) - to read this code, literally makes me cringe - if i were to maintain this code-base, the very first thing i would do, is to normalize all of the white-space

o/c i am not maintaining this code; so i am happy to drop this discussion of tidiness - i just would not suggest any more housekeeping tasks in the future - o/c there is plenty of significant work awaiting an author - i just tend to want to have things tidy before i begin anything important - i can grin and bear the parts i dont like - i just wont work on those parts unless i can clean them up first (is all im getting at)

Reply:
ok i will indulge :)

most of the files do not end with a newline - when viewing the file with `cat`
for example, the next shell prompt will begin offset at the end of the last line

```
$ printf text > deleteme ; cat deleteme ;
text$ _
```

git and other tools such as tig even warn about that - my editor (kate)
automatically inserts an ending newline, if it is missing

```
$ printf text > deleteme ; git add deleteme ; git commit -m deleteme deleteme ;
$ git diff HEAD^
diff --git a/deleteme b/deleteme
new file mode 100644
index 00000000..f3a34851
--- /dev/null
+++ b/deleteme
@@ -0,0 +1 @@
+text
\ No newline at end of file
```

my editor also removes all trailing white-space automatically - this code-base
has many of those - o/c i could disable that feature, but instead, i usually
avoid committing those white-space change to any feature commit, but collect
them into a 'housekeeping' commit

number of files with lines ending in white-space
```
$ grep -rlIE  '\s$' | wc -l
329
```
number of lines ending in white-space
$ grep -rIE  '\s$' | wc -l
3518

the number of spaces that a tab resolves to is less important if tabs and
spaces are never mixed; but the default for many editors is make tabs 8 chars
wide, which makes code which uses tabs for anything other than monotonically
increasing scope indentation look very wonky

very long lines (>100 chars)also make it difficult to read, depending on the
reader's font size, or how some forge presents the file - people have argued
about this for years o/c - IMHO, 80 cols is a good "lowest common denominator"
- i usually aim for 80 cols, but allow myself to go a bit over 90 - i think 120
is beyond excessive (as the example below)

the main Makefile exhibits most of these white-space issues - this is how
some selected parts of that file looks in tig:

> DOCUMENTS =  $(DOCDIR)*.pdf $(DOCDIR)*.txt $(BINDIR)neo6502.inc
> BINARIES =      $(BINDIR)*.uf2 $(BINDIR)*.elf $(ROOTDIR)emulator$(S)cross-compile$(S)neowin.zip $(BINDIR)basic.bin \
>                         $(ROOTDIR)emulator$(S)neolinux.zip
> PYTHONAPPS = $(BINDIR)makebasic.zip $(BINDIR)listbasic.zip $(BINDIR)createblanks.zip $(BINDIR)makeimg.zip \
>                         $(BINDIR)nxmit.zip

those three lines have a mixture of tabs and spaces - those will appear as the author sees
them, only if each reader has tabs the same width, staggered as above otherwise

> # ***************************************************************************************
> #
> #                                                              Make the release zip
> #
> # ***************************************************************************************

when i see that, i ask myself, "why should a sentence as brief as 'Make the release zip'
need to wrap on an 80 column display"?

Reply:
Okay, makes a lot of sense. Is there a tool which will do this conversion ? Tabs->Spaces and right strip and so on.  Something we could use on the entire source code base ? 

Probably stuck with 4 space indents, but I have no problem with this. I see what your problem is. From my own POV it is easy enough to change the editor to use spaces.

Reply:
blindly replacing all tabs with spaces, easy-peasy - i would use `sed` - but most editors also have that feature

i tried - ideally, it could be as simple as this one-liner:

```
$ sed -i 's|\t|    |' $(git ls-files | grep -vE 'Makefile$')
```

that misses two files though; because they have spaces in the filenames - i would rename those - that helps a lot with shell scripting - in order to deal with filenames with spaces, the command would need to be more complicated

```
$ git ls-files -z | grep -vzE 'Makefile$' | xargs -0 sed -i 's|\t|    |'
```

i probably would not do that though, unless i was going to scrutinize the diff - it is still error-prone to blindly replace all tabs - for example, that is why `grep -v Makefile` - Makefile rules require tabs - if white-space has semantics in any other files (eg: python), it could break those

Reply:
I asked for advice on discord and a tool was suggested, so I'll close this and adopt your changes as suggested. Many thanks.